scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala

```diff
@@ -6,83 +6,149 @@ import io.circe.generic.auto.*
+          }
+      }
 
-      (folderInfoOfEntitiesThatDoNotExist, folderInfoOfEntitiesThatExist) = folderInfoWithExpectedParentRef.partition(
-        _.entity.isEmpty
-      )
-      _ <- createFolders(folderInfoOfEntitiesThatDoNotExist, dependencies.entityClient, secretName)
-      _ <- logWithBatchRef(s"Created ${folderInfoOfEntitiesThatDoNotExist.length} entities which did not previously exist")
-      _ <- verifyEntitiesAreStructuralObjects(folderInfoOfEntitiesThatExist)
-
-      folderInfoOfEntitiesThatExistWithSecurityTags <-
-        verifyExpectedParentFolderMatchesFolderFromApiAndGetSecurityTag(folderInfoOfEntitiesThatExist)
-      folderUpdateRequests = findOnlyFoldersThatNeedUpdatingAndCreateRequests(
-        folderInfoOfEntitiesThatExistWithSecurityTags
+    def addFolder(folderRow: ArchiveFolderDynamoTable, entityMap: Map[UUID, UUID]) = {
+      val potentialParentRef = for {
+        directParent <- folderRow.directParentId
+        entityRef <- entityMap.get(directParent)
+      } yield entityRef
+      val addFolderRequest = AddEntityRequest(
+        None,
+        folderRow.potentialTitle.getOrElse(folderRow.name),
+        folderRow.potentialDescription,
+        StructuralObject,
+        Open,
+        potentialParentRef
       )
+      val folderName = folderRow.name
+      val identifiersToAdd = List(Identifier(sourceId, folderName)) ++
+        folderRow.identifiers.map(id => Identifier(id.identifierName, id.value))
+      for {
+        entityId <- dependencies.entityClient.addEntity(addFolderRequest)
+        _ <- identifiersToAdd.traverse { identifierToAdd =>
+          dependencies.entityClient.addIdentifierForEntity(
+            entityId,
+            StructuralObject,
+            identifierToAdd
+          )
+        }
+      } yield entityId
+    }
+
+    def getEntitiesByIdentifier(
+        folderRows: List[DynamoTable]
+    ): IO[Map[String, Entity]] = {
+      folderRows
+        .map(_.name)
+        .toSet
+        .map(name => Identifier(sourceId, name))
+        .toList
+        .traverse(identifier =>
+          dependencies.entityClient.entitiesByIdentifier(identifier).flatMap {
+            case head :: next =>
+              if next.nonEmpty then IO.raiseError(new Exception(s"There is more than 1 entity with the same SourceID ${identifier.value}"))
+              else IO(identifier.value -> Option(head))
+            case Nil => IO(identifier.value -> None)
+          }
+        )
+        .map {
+          _.collect { case (value, Some(entity)) =>
+            (value, entity)
+          }.toMap
+        }
+    }
 
-      _ <- folderInfoOfEntitiesThatExistWithSecurityTags.map { fi =>
-        val identifiersFromDynamo = fi.folderRow.identifiers.map(id => Identifier(id.identifierName, id.value))
+    val logWithBatchRef = log(Map("batchRef" -> stepFnInput.batchId))(_)
 
-        val entity = fi.entity.get
+    for {
+      folderRows <- getFolderRows
```

Comment:
It's a bit long again and it is used quite a bit so it makes it quite hard to parse. I've added an implicit ordering to the method definition so you can see it's sorting something.

Reply:
Ok. Maybe we could also give the `given` a name?

Reply:
It's not really necessary so I'll leave it. As I say, I'd like to move a bit more towards using the types to see what a method is doing rather than naming everything explicitly